### ISSUE TITLE
Do not show notifications if the app is active

### ISSUE THREAD

**Issue by borisyankov** [zulip/zulip-mobile#3114]
If the app is on-screen, currently being used, do not show notifications!

**Comment by kunall17** [comment]
We can use the setting `notification when online` for this one?

**Comment by chrisbobbe** [comment]
@gnprice @ray-kraesig, are you aware of notifications showing up when the app is active?

Note that this is distinct from https://github.com/zulip/zulip-mobile/issues/3119, which reports that we can reduce the volume of notifications by removing those for messages we've read on the web.

**Comment by gnprice** [comment]
Chris and I experimented a bit (with his iPhone, on chat.zulip.org) and determined that the behavior here isn't quite anything I thought it was. :-)

In particular we didn't actually succeed in reproducing getting a notification while the app was on the screen. This even after he set "Send mobile notificatinos even if I'm online" to true.

So the first step in doing something here will be to try to pin down what the existing pattern of behavior is for how notifications interact with the app already being open.

Stepping back to the original description / request: I think it's actually important that there *should* be a notification when you get a (notifiable) message, even when you're actively using the app -- if you're using it in some other narrow, or anywhere else that means you're not going to immediately see the message in the app. We wouldn't want it to be the case that when you're actively using the app, that causes you to not see PMs unless you keep actively checking for them.

But I do think it's a not-great experience if you're actively participating in a PM thread, and you keep getting a ding and a banner for every message the other / another person sends in the thread. The ideal behavior here might be: when you're looking at a message list, suppress notifications for messages that would appear in that same message list.

**Comment by timabbott** [comment]
Yeah, I believe that's the right UX design.

**Comment by timabbott** [comment]
Just to extend that commentary, what we do in the webapp is we suppress notifications for messages that appear onscreen; which is a bit different from being in the message list (but I think matches the user expectation for whether it's annoying or useful to get a ding).

**Comment by gnprice** [comment]
Yeah, that makes sense. That might be the ideal behavior for the mobile app, too.

One difference from web is that we have the "N unreads" banner, which is prominent at the top of the screen (in the same general area a notification would appear, in fact) in a way that the counts in the left sidebar aren't. But, when I think about situations where this would arise -- like I'm catching up on a long thread, and someone @-mentions me in it, down below what's onscreen -- I feel like I do want the notification.

**Comment by gnprice** [comment]
Bump to mention that we've heard a complaint about this [on czo](https://chat.zulip.org/#narrow/stream/48-mobile/topic/PM.20notifications.20Even.20if.20I'm.20chatting/near/918309).

**Comment by gnprice** [comment]
The same user who highlighted this issue for us a few months ago has [provided a video](https://chat.zulip.org/#narrow/stream/48-mobile/topic/PM.20notifications.20Even.20if.20I'm.20chatting/near/1013931). It shows that the behavior they're currently seeing is the same one originally reported in this issue, unlike what we saw in experiments described above at https://github.com/zulip/zulip-mobile/issues/3114#issuecomment-585966226 .

Specifically:
* they're looking at a PM conversation narrow
* they get a message in that conversation; it appears in the message list, and gets marked as read
* then a few moments later, a notification appears.

I suspect that what we're seeing is an OS difference. That video shows an Android device; the experiment above was on iOS. So I suspect the situation is:
* On Android, we're showing notifications even when the app is in the foreground (and even when in fact the very same message appears on-screen in the app.)
* On iOS, we're not showing notifications when the app is in the foreground at all (regardless of whether the app is currently showing that conversation or only some entirely unrelated conversation.)

The behavior we want is a mix of these, where whether we show a notification depends on whether the message will already be shown on-screen. This will require implementing some new logic in order to be able to use that information when determining what notifications to show.

**Comment by chrisbobbe** [comment]
> I suspect that what we're seeing is an OS difference.

I agree.

> This will require implementing some new logic in order to be able to use that information when determining what notifications to show.

Correct; I've got a few bullet points to brainstorm at https://github.com/zulip/zulip-mobile/issues/4241#issue-686854636.

**Comment by andreasoc** [comment]
any update?

**Comment by chrisbobbe** [comment]
No, sorry, @andreasoc. Progress on iOS notification work is stalled on https://github.com/zulip/zulip-mobile/issues/4115#issuecomment-792887302.

**Comment by andreasoc** [comment]
[CZO discussion](https://chat.zulip.org/#narrow/stream/48-mobile/topic/UI.20Compose.20Experience.20to.20Improve) with some details
